bullet/src/SDFFeatures.cc

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "SDFFeatures.hh"
+#include <ignition/math/eigen3/Conversions.hh>
+#include <ignition/math/Helpers.hh>
+
+#include <sdf/Geometry.hh>
+#include <sdf/Box.hh>
+#include <sdf/Sphere.hh>
+#include <sdf/Cylinder.hh>
+#include <sdf/Plane.hh>
+#include <sdf/JointAxis.hh>
+
+namespace ignition {
+namespace physics {
+namespace bullet {
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfWorld(
+    const Identity &_engine,
+    const ::sdf::World &_sdfWorld)
+{
+  const Identity worldID = this->ConstructEmptyWorld(_engine, _sdfWorld.Name());
+
+  const WorldInfoPtr &worldInfo = this->worlds.at(worldID);
+
+  auto gravity = _sdfWorld.Gravity();
+  worldInfo->world->setGravity(btVector3(gravity[0], gravity[1], gravity[2]));
+  return worldID;
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfModel(
+    const Identity &_worldID,
+    const ::sdf::Model &_sdfModel)
+{
+  // Read sdf params
+  const std::string name = _sdfModel.Name();
+  const math::Pose3d pose = _sdfModel.RawPose();
+  const bool isStatic = _sdfModel.Static();
+  // Links within a model will collide unless these are chained with a joint
+  // const bool selfCollide = _sdfModel.SelfCollide();
+
+  const auto modelIdentity = this->AddModel({name, _worldID, isStatic, pose});
+
+  // After creating all the links, join the ones that have joints
+  for (std::size_t i=0; i < _sdfModel.JointCount(); ++i)
+  {
+    igndbg << "Loop adding joints.\n";
+
+    const ::sdf::Joint *sdfJoint = _sdfModel.JointByIndex(i);
+    if (!sdfJoint)
+    {
+      ignerr << "The joint with index [" << i << "] in model ["
+             << _sdfModel.Name() << "] is a nullptr. It will be skipped.\n";
+      continue;
+    }
+  }
+
+  return modelIdentity;
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfLink(
+    const Identity &_modelID,
+    const ::sdf::Link &_sdfLink)
+{
+  // Read sdf params
+  const std::string name = _sdfLink.Name();
+  const math::Pose3d pose = _sdfLink.RawPose();
+  const ignition::math::Inertiald inertial = _sdfLink.Inertial();
+  auto mass = inertial.MassMatrix().Mass();
+  const auto diagonalMoments = inertial.MassMatrix().DiagonalMoments();
```

Comment:
This only takes the diagonal moments the mass matrix and ignores the off diagonal terms. Based on the Gazebo-classic implementation, If I understand correctly, when constructing `btRigidBody`, we have to give bullet 3 quantities: 
 1. the position of the body frame. The origin of this frame is a the center of mass of the body.
 2. the orientation of the body frame such that the inertia matrix is diagonal. This means we have to diagonalize the inertia matrix and apply the resulting rotation offset to the pose of the body frame.
 3. the diagonal moments of the diagonalized inertia matrix.

To accomplish (1), we have to compose the model pose, the link pose and the inertia pose. I think `poseIsometry` is missing `inertial.Pose()`.

For (2), we have to first use `ignition::math::MassMatrix3::PrincipalMoments` which diagonalizes the inertia matrix. `ignition::math::MassMatrix3::DiagonalMoments` simply gives the values on the diagonal of the matrix.

After that, we have to obtain the rotation offset using `PrincipalAxesOffset` and apply that to `inertial.Pose()`.

Reply:
https://github.com/ignitionrobotics/ign-physics/pull/226 should take care of this. @azeey If you can think of an easy way to test it let me know. I can't wrap my head around the usage of the inertial pose yet